### PR TITLE
[REF] *: Update Chart.js to V4.3

### PR DESCRIPTION
**Pull request by pipu-odoo**
Change version of Chart.js from V2.9.3 to V4.3.0
You can find changelog on  
https://www.chartjs.org/docs/latest/migration/v4-migration.html

Version 4.3.0 of chart.js no longer uses moment.js. This task is a
prerequisite to completely remove the use of moment.js (task 3391739)

Why add aspectRatio : 2 ? => Canvas aspect ratio (i.e. width / height, 
a value of 1 representing a square canvas). Note that this option
 is ignored if the height is explicitly defined either as attribute 
or via the style. The default value varies by chart type; Radial charts 
(doughnut, pie, polarArea, radar) default to 1 and others default to 2.

Why no more Chart.animationService.advance(); ? 
There is no longer an equivalent in this version. 
However, we have verified that the problem is no longer present in this
version.

Why use now getElementsAtEventForMode ? 
It's clearly noted in the changelog. follow the link above.

task-3392075

### CODE REVIEW

File: addons/survey/static/src/js/survey_result.js [hunk 1]
```diff
@@ -325,20 +332,33 @@ publicWidget.registry.SurveyResultChart = publicWidget.Widget.extend({
...
+                    title: {
+                        display: true,
+                        text: _t("Performance by Section"),
+                    },
+                    legend: {
+                        display: true,
+                    },
+                    tooltip: {
+                        callbacks: {
+                            label: function (tooltipItem, data) {
```
Comment (LucasLefevre) [inline addons/survey/static/src/js/survey_result.js:345]: outdated API.
The new API is `function (tooltipItem)`
https://www.chartjs.org/docs/latest/api/interfaces/TooltipCallbacks.html#label

Same issue in the enterprise branch in `hr_payroll` (doesn't seem to have a PR yet)

https://github.com/odoo-dev/enterprise/blob/e5d4b0d5634268ac8e4c6af1d392eaa403c1b3be/hr_payroll/static/src/components/dashboard/payroll_stats/payroll_stats.js#L249

File: addons/web/static/lib/Chart/Gauge.js
Comment (LucasLefevre) [inline addons/web/static/lib/Chart/Gauge.js:1]: this file can be removed since it's implemented directly in o-spreadsheet

Comment (pipu-odoo) [reply addons/web/static/lib/Chart/Gauge.js:1]: okay

File: addons/website/static/src/snippets/s_chart/000.js [hunk 2]
```diff
@@ -8,6 +8,7 @@ const ChartWidget = publicWidget.Widget.extend({
     disabledInEditableMode: false,
     jsLibs: [
         '/web/static/lib/Chart/Chart.js',
+        '/web/static/lib/chartjs-adapter-luxon/chartjs-adapter-luxon.js',
```
Comment (rdeodoo) [inline addons/website/static/src/snippets/s_chart/000.js:11]: Something is making this snippet crash
image

- Go to website app
- Enter edit mode
- Drag & Drop "Chart" snippet
- Change "type" to something else
- Crash

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/spreadsheet/static/lib/chartjs-gauge/chartjs-gauge.js
Comment (LucasLefevre) [inline addons/spreadsheet/static/lib/chartjs-gauge/chartjs-gauge.js:1]: this file can be removed

File: addons/survey/static/src/js/survey_result.js [hunk 3]
```diff
@@ -146,17 +147,19 @@ publicWidget.registry.SurveyResultChart = publicWidget.Widget.extend({
...
+                    x: {
+                        axis: "x",
                         ticks: {
                             callback: this._customTick(25),
                         },
-                    }],
-                    yAxes: [{
+                    },
+                    y: {
+                        axis: "y",
```
Comment (LucasLefevre) [inline addons/survey/static/src/js/survey_result.js:157]: some useless `axis: "x"` and `axis: "y"` all around

File: addons/web/static/tests/views/graph_view_tests.js [hunk 4]
```diff
@@ -1050,7 +1050,7 @@ QUnit.module("Views", (hooks) => {
         checkModeIs(assert, graph, "line");
         assert.strictEqual(graph.model.metaData.stacked, true, "graph should be stacked.");
         assert.strictEqual(
-            getScaleY(graph).every((y) => y.stacked),
+            [getScaleY(graph)].every((y) => y.stacked),
```
Comment (LucasLefevre) [inline addons/web/static/tests/views/graph_view_tests.js:1053]:
```suggestion
            getScaleY(graph).stacked),
```

File: addons/survey/static/src/js/survey_result.js (hunk 3 above)
Comment (pipu-odoo) [reply addons/survey/static/src/js/survey_result.js:157]: Not so useless if you use it to grep all axis for chart.js

File: addons/website/static/src/snippets/s_chart/000.js [hunk 5]
```diff
@@ -63,6 +65,9 @@ const ChartWidget = publicWidget.Widget.extend({
                     beginAtZero: true,
                 }
             };
+        } else if (this.el.dataset.type === 'horizontalBar') {
```
Comment (pipu-odoo) [inline addons/website/static/src/snippets/s_chart/000.js:68]: @rdeodoo Fixed the bug

**Review by rdeodoo — CHANGES_REQUESTED** [review]
> @rdeodoo Fixed the bug

Well you can now change the `Type` option without the JS crashing but the `Legend` option just below is not working anymore with the lib change.

Once this is adapted, @pierre-pde can have a round of test if you are still unsure everything works fine.

image

**Comment by pierre-pde** [comment]
@pipu-odoo , as discussed on Discord, there is a size issue with graph size for type "pie";"doughnut" et "radar" 
Video comparing master (beginning) with this branch (second part of video) https://bit.ly/445TUlz

**Comment by pierre-pde** [comment]
@pipu-odoo , looks good now, thanks!

**Review by rdeodoo — APPROVED** [review]
Approving for the website part

**Review by LucasLefevre — COMMENTED** [review]
robodoo rebase-merge

**Review by brboi — COMMENTED** [review]
I only reviewed the first commit.

I understand that most of the diff comes from simple API changes.

The comments mostly consist of nitpicking or to ask more information, in order to document the whole PR.

Regarding the commit message, could you:
- specify the full version 4.3.0
- specify the version being replaced
- add a link to the changelog between those two versions
- maybe also note some of the outstanding changes between those two versions

**Review by LucasLefevre — APPROVED** [review]
robodoo r+
Thanks :)

**Comment by flch-odoo** [comment]
I wouldn't do that, problems in Survey are not solved. See https://49277096-master-all.runbot124.odoo.com/survey/results/mycompany-vendor-certification-2

**Comment by rdeodoo** [comment]
@robodoo r-

**Comment by rdeodoo** [comment]
Website link is also broken, checked out of curiosity..
image

Please, test your development more intensively.
This PR is basically updating a library and apart of spreadsheet, it's adapting:
- website chart snippet
- survey
- website links

3/3 are not correctly adapted, the chart snippet was even fully broken (traceback and not usable anymore, breaking full page edition) and took a few attempts to be correctly fixed while it was just a matter of comparing side to side with regular master.

**Review by LucasLefevre — COMMENTED** [review]
robodoo r-
some broken stuff in survey (see task pad)

**Comment by pipu-odoo** [comment]
I made adaptations based on visual comparisons for surveys (checked with flch) and website links.
Gauge field, graph renderer and website were already done. 
@rdeodoo : as your screenshot was not very clear about where you made it, could you confirm that the graphic is now correct ?

**Review by rdeodoo — APPROVED** [review]
Validating the website & website links part

**Comment by LucasLefevre** [comment]
robodoo r+

**Comment by LucasLefevre** [comment]
robodoo rebase-merge

**Pull request closed without merging**